### PR TITLE
Support for time datatype

### PR DESCRIPTION
**Pull request by mbudiu-vmw**
If this is good I will do Date in a similar way.

### CODE REVIEW

File: lib/time.dl [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+extern function try_from_hms_nano(hours: bit<8>, minutes: bit<8>, seconds: bit<8>, microseconds: bit<32>): Result<time, string>
+extern function midnight(): time
+extern function hour(t: time): bit<8>
+extern function minute(t: time): bit<8>
+extern function second(t: time): bit<8>
+extern function millisecond(t: time): bit<16>
+extern function microsecond(t: time): bit<32>
+extern function nanosecond(t: time): bit<32>
+extern function time2string(t: time): string
+extern function parse(s: string, format: string): Result<time, string>
```
Comment (ryzhyk) [inline lib/time.dl:15]: Can you add examples of supported input formats?

File: lib/time.dl [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+extern function try_from_hms_micro(hours: bit<8>, minutes: bit<8>, seconds: bit<8>, microseconds: bit<32>): Result<time, string>
+extern function try_from_hms_nano(hours: bit<8>, minutes: bit<8>, seconds: bit<8>, microseconds: bit<32>): Result<time, string>
+extern function midnight(): time
+extern function hour(t: time): bit<8>
+extern function minute(t: time): bit<8>
+extern function second(t: time): bit<8>
+extern function millisecond(t: time): bit<16>
+extern function microsecond(t: time): bit<32>
+extern function nanosecond(t: time): bit<32>
+extern function time2string(t: time): string
```
Comment (ryzhyk) [inline lib/time.dl:14]: Can we also have a version of `2string` that takes format string?

File: lib/time.dl [hunk 3]
```diff
@@ -0,0 +1,15 @@
+extern type time
+
+extern function try_from_hms(hours: bit<8>, minutes: bit<8>, seconds: bit<8>): Result<time, string>
```
Comment (ryzhyk) [inline lib/time.dl:3]: I would prefix all these functions with `time_`. Alternatively, users can import this module using qualified name `import time as time`, but then what happens when you start adding support for dates and date-times?

File: test/datalog_tests/time_test.dat [hunk 4]
```diff
@@ -0,0 +1,4 @@
+start;
```
[the remark this answers is not in the document]
Comment (mbudiu-vmw) [reply test/datalog_tests/time_test.dat:1]: Flatbuf is not supported. It's not clear what needs to be changed to support extern types in flatbuf.

[the remark this answers is not in the document]
Comment (ryzhyk) [reply test/datalog_tests/time_test.dat:1]: CLI does not use flatbuf.

File: lib/time.dl (hunk 3 above)
Comment (mbudiu-vmw) [reply lib/time.dl]: I've done everything but add this prefix.

File: rust/template/src/main.rs [hunk 5]
```diff
@@ -95,7 +94,7 @@ fn handle_cmd(
         Command::Comment => Ok(()),
         Command::Rollback => hddlog.transaction_rollback(),
         Command::Timestamp => {
-            println!("Timestamp: {}", precise_time_ns());
+            println!("Timestamp: {:#?}", std::time::Instant::now());
```
Comment (ryzhyk) [inline rust/template/src/main.rs:97]: why did you have to change this. The new method will produce output that does not look good, e.g., `Instant { tv_sec: 5122222, tv_nsec: 893024142 }`

**Pull request merged**